library("Biostrings")
library("dplyr")
library("stringr")

# example  megacc.exe -a muscle_align_nucleotide.mao -d Cronobacter_OACs.fasta -o align
# example  megacc.exe -a infer_ML_nucleotide -d align.meg -o tree
mega <- function(align = "megacc.exe", parameter, infile, outfile) {
  system2(align,
          args = c("-a", parameter,
                   "-d", infile,
                   "-o", outfile))
}


# execute local-installed makeblastdb program
# example
makeblast_database <- function(makeblastdb = "makeblastdb", infile, outfile, dbtype = "nucl") {
  system2(makeblastdb,
          args = c("-in", infile,
                   "-dbtype", dbtype,
                   "-out", outfile))
}


# execute local-installed blastn program
# example
blast <- function(blast="blastn", query, database, outfmt = 6, outfile = "test.txt") {
  system2(blast,
          args = c("-query", query,
                   "-db", database,
                   "-outfmt", outfmt,
                   "-out", outfile))
}

# transform name of sequence in sequence_file
# require Biostrings::readDNAStringSet() function
tran_name <- function(sequence_file) {
  seq <- readDNAStringSet(sequence_file)
  new_nm <- c()
  for(i in names(seq)) {
    m = unlist(strsplit(i, split = " "))[1]
    new_nm <- c(new_nm, m)
  }
  names(seq) <- new_nm
  return(seq)
}


# get genome length from genome sequence
genome_length <- function(sequence) {
  len <- sum(sequence@ranges@width)
  return(len)
}


# extract certain contig length from genome sequence
contig_length <- function(sequence, pair_name){
  for(i in seq(length(sequence))) {
    if(sequence@ranges@NAMES[i] == pair_name) {
      return(sequence@ranges@width[i])
    }
  }
}

extract_len <- function(sequence, pair_name){
  for(i in seq(length(sequence))) {
    if(sequence@ranges@NAMES[i] == pair_name) {
      return(sequence@ranges@width[i])
    }
  }
}

# extract genome name from genome (去除路径信息以及后缀)
extract_name <- function(genome) {
  genome_name <- str_split(genome, "/")[[1]][length(str_split(genome, "/")[[1]])]
  if(endsWith(genome_name, ".fasta")) {
    genome_name <- str_sub(genome_name, end = nchar(genome_name)-6)
  }
  else if(any(endsWith(genome_name, c(".fna",".fas")))) {
    genome_name <- str_sub(genome_name, end = nchar(genome_name)-4)
  }
  return(genome_name)
}



generate_blast <- function(sequence, genome, gene) {
  gene_file <- extract_seq(sequence = sequence, pair_name = gene, type = 0)
  new_file <- paste0(gene, ".fasta")
  writeLines(gene_file, con = new_file, sep = "\n")
  makeblast_database(infile = new_file, outfile = paste0(gene, "_db"))
  out_file <- paste0(gene, ".txt")
  blast(query = genome, database = paste0(gene, "_db"), outfile = out_file)
  if(length(readLines(out_file)) == 0) {
    return("bad")
  }
  else {
    blast_result <- readLines(out_file, n=1)
    unlink(c(new_file, paste0(gene,"_db",".nin"), paste0(gene,"_db",".nhr"),
             paste0(gene,"_db",".nsq"), out_file), force = TRUE, recursive = FALSE)
    return(blast_result)
  }
}


# extract specie from tree generated by mega
extract_specie <- function(a, b) {
  seven_species <- c("condimenti","dublinensis","malonaticus","muytjensii","sakazakii","turicensis","universalis")
  locate <- as.vector(str_locate(a,b))
  left_bra <- str_locate_all(a, "\\(")[[1]][,1]
  right_bra <- str_locate_all(a, "\\)")[[1]][,1]
  d1 <- left_bra[max(which(left_bra < locate[1]))]
  d2 <- right_bra[min(which(right_bra > locate[1]))]
  m <- substr(a, start = d1+1, stop = d2-1)
  m1 <- unlist(strsplit(m, ","))
  if (length(m1) == 2) {
    n1 <- unlist(strsplit(m1[1], ":"))[1]
    n2 <- unlist(strsplit(m1[2], ":"))[1]
    if(n1 == b) {
      for(i in seven_species) {
        if(i == n2) return(n2)
      }
    }
    else if (n2 == b) {
      for(i in seven_species) {
        if(i == n1) return(n1)
      }
    }
  }
  else {
    return("bad")
  }
}

